File: src/spessasynth_lib/synthetizer/worklet_system/cppessa_synth/cpessasynth_components/unit_converter/UnitConverter.cpp
//
// Created by spessasus on 21.05.24.
//

#include <cmath>
#include <emscripten/emscripten.h>
#include <cstdio>
#include "UnitConverter.h"

float UnitConverter::timecentLookupTable[MAX_TIMECENT - MIN_TIMECENT + 1];
float UnitConverter::absoluteCentLookupTable[MAX_ABS_CENT - MIN_ABS_CENT + 1];
//float UnitConverter::decibelLookupTable[(MAX_DECIBELS - MIN_DECIBELS) * DECIBEL_TABLE_STATIC_POINT + 1];

EMSCRIPTEN_KEEPALIVE
void UnitConverter::initializeLookupTables() {
    // timecent
    for (int i = 0; i < MAX_TIMECENT - MIN_TIMECENT + 1; ++i) {
        int timecents = i + MIN_TIMECENT;
        UnitConverter::timecentLookupTable[i] = powf(2.0f, (float)timecents / 1200.0f);
    }
    printf("\n");

    // absolute cent
    for (int i = 0; i < MAX_ABS_CENT - MIN_ABS_CENT + 1; ++i) {
        int absoluteCents = i + MIN_ABS_CENT;
        UnitConverter::absoluteCentLookupTable[i] = 440.0f * powf(2.0f, ((float)absoluteCents - 6900.0f) / 1200.0f);
    }

    // decibel (2 points of precision)
//    for (int i = 0; i < (MAX_DECIBELS - MIN_DECIBELS) * 100 + 1; ++i) {
//        float decibels = (MIN_DECIBELS * DECIBEL_TABLE_STATIC_POINT + (float)i) / DECIBEL_TABLE_STATIC_POINT;
//        UnitConverter::decibelLookupTable[i] = powf(10.0f, -decibels / 20.0f);
//        printf("%f %f ", UnitConverter::decibelLookupTable[i], decibels);
//    }
//    printf("\n");
}

EMSCRIPTEN_KEEPALIVE
float UnitConverter::timecentsToSeconds(int timecents) {
    return UnitConverter::timecentLookupTable[timecents - MIN_TIMECENT];
}

EMSCRIPTEN_KEEPALIVE
float UnitConverter::absCentsToHz(int absoluteCents) {
    return UnitConverter::absoluteCentLookupTable[absoluteCents - MIN_ABS_CENT];
}

EMSCRIPTEN_KEEPALIVE
float UnitConverter::decibelAttenuationToGain(float decibels) {
    return powf(10.0f, -decibels / 20.0f); //UnitConverter::decibelLookupTable[(unsigned int)floor((decibels - MIN_DECIBELS) * DECIBEL_TABLE_STATIC_POINT)];
}

File: src/spessasynth_lib/synthetizer/audio_engine/engine_components/unit_converter.js
/**
 * unit_converter.js
 * purpose: converts soundfont units into more useable values with the use of lookup tables to improve performance
 */


// timecent lookup table
const MIN_TIMECENT = -15000;
const MAX_TIMECENT = 15000;
const timecentLookupTable = new Float32Array(MAX_TIMECENT - MIN_TIMECENT + 1);
for (let i = 0; i < timecentLookupTable.length; i++)
{
    const timecents = MIN_TIMECENT + i;
    timecentLookupTable[i] = Math.pow(2, timecents / 1200);
}

/**
 * Converts timecents to seconds
 * @param timecents {number} timecents
 * @returns {number} seconds
 */
export function timecentsToSeconds(timecents)
{
    if (timecents <= -32767)
    {
        return 0;
    }
    return timecentLookupTable[timecents - MIN_TIMECENT];
}

// abs cent lookup table
const MIN_ABS_CENT = -20000; // freqVibLfo
const MAX_ABS_CENT = 16500; // filterFc
const absoluteCentLookupTable = new Float32Array(MAX_ABS_CENT - MIN_ABS_CENT + 1);
for (let i = 0; i < absoluteCentLookupTable.length; i++)
{
    const absoluteCents = MIN_ABS_CENT + i;
    absoluteCentLookupTable[i] = 440 * Math.pow(2, (absoluteCents - 6900) / 1200);
}

/**
 * Converts absolute cents to hertz
 * @param cents {number} absolute cents
 * @returns {number} hertz
 */
export function absCentsToHz(cents)
{
    if (cents < MIN_ABS_CENT || cents > MAX_ABS_CENT)
    {
        return 440 * Math.pow(2, (cents - 6900) / 1200);
    }
    return absoluteCentLookupTable[~~(cents) - MIN_ABS_CENT];
}

// decibel lookup table (2 points of precision)
const MIN_DECIBELS = -1660;
const MAX_DECIBELS = 1600;
const decibelLookUpTable = new Float32Array((MAX_DECIBELS - MIN_DECIBELS) * 100 + 1);
for (let i = 0; i < decibelLookUpTable.length; i++)
{
    const decibels = (MIN_DECIBELS * 100 + i) / 100;
    decibelLookUpTable[i] = Math.pow(10, -decibels / 20);
}

/**
 * convers decibel attenuation to gain
 * @param decibels {number} the decibel attenuation
 * @returns {number} gain
 */
export function decibelAttenuationToGain(decibels)
{
    return decibelLookUpTable[Math.floor((decibels - MIN_DECIBELS) * 100)];
}